### PR TITLE
feat(doctrine): add new filter for filtering an entity using PHP backed enum, resolves #6506

### LINKED ISSUE #6506: Support for BackedEnumFilter
**Description**  

Is a new Doctrine filter BackedEnumFilter could be interesting for filtering an entity using backed enum?

**Example**  

Could be used in this way:
```php
<?php

namespace App\Entity;

#[ORM\Entity(repositoryClass: UserRepository::class)]
#[ApiResource(
    operations: [
        new Get(normalizationContext: ['groups' => 'user:item']),
        new GetCollection(normalizationContext: ['groups' => 'user:list']),
    ],
)]
#[ApiFilter(BackedEnumFilter::class, properties: ['status'])]
class User
{
    #[ORM\Column(enumType: UserStatus::class, options: ['default' => UserStatus::Active])]
    #[Groups(['user:list', 'user:item'])]
    private UserStatus $status = UserStatus::Active;
}
```

with an Enum like this:
```php
<?php

enum UserStatus: string
{
    case Active = 'active';
    case Blocked = 'blocked';
}
```

### PR DESCRIPTION
**Pull request by mremi**
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | Closes #6506
| License       | MIT
| Doc PR        | api-platform/docs#... 

Adding new Doctrine filter for filtering an entity using PHP backed enum

### CODE REVIEW

File: src/Doctrine/Common/Filter/BackedEnumFilterTrait.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    abstract protected function normalizePropertyName(string $property): string;
+
+    /**
+     * Determines whether the given property refers to a backed enum field.
+     */
+    abstract protected function isBackedEnumField(string $property, string $resourceClass): bool;
+
+    private function normalizeValue($value, string $property): mixed
+    {
+        if (false !== array_search($value, array_map(fn (\BackedEnum $case) => $case->value, $this->enumTypes[$property]::cases()), true)) {
```
Comment (soyuka) [inline src/Doctrine/Common/Filter/BackedEnumFilterTrait.php:83]:
```suggestion
        $values = array_map(fn (\BackedEnum $case) => $case->value, $this->enumTypes[$property]::cases()), true);
        if (in_array($value, $values)) {
```

File: src/Doctrine/Common/Filter/BackedEnumFilterTrait.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            return $value;
+        }
+
+        $this->getLogger()->notice('Invalid filter ignored', [
+            'exception' => new InvalidArgumentException(\sprintf('Invalid backed enum value for "%s" property, expected one of ( "%s" )',
+                $property,
+                implode('" | "', array_map(fn (\BackedEnum $case) => $case->value, $this->enumTypes[$property]::cases())))
```
Comment (soyuka) [inline src/Doctrine/Common/Filter/BackedEnumFilterTrait.php:90]:
```suggestion
                implode('" | "', $values)
```

File: src/Doctrine/Orm/Filter/BackedEnumFilter.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+        if (!$metadata instanceof ClassMetadata) {
+            return false;
+        }
+
+        $fieldMapping = $metadata->fieldMappings[$propertyParts['field']];
+
+        // Doctrine ORM 2.x returns an array and Doctrine ORM 3.x returns a FieldMapping object
+        if ($fieldMapping instanceof FieldMapping) {
+            $fieldMapping = (array) $fieldMapping;
```
Comment (soyuka) [inline src/Doctrine/Orm/Filter/BackedEnumFilter.php:163]: does `main` support ORM 2.x ? I think that we can add the code only for `3.x`

File: src/Doctrine/Orm/Tests/Filter/BackedEnumFilterTest.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+namespace ApiPlatform\Doctrine\Orm\Tests\Filter;
+
+use ApiPlatform\Doctrine\Orm\Filter\BackedEnumFilter;
+use ApiPlatform\Doctrine\Orm\Tests\DoctrineOrmFilterTestCase;
+use ApiPlatform\Doctrine\Orm\Tests\Fixtures\Entity\Dummy;
+
+/**
+ * @author Rémi Marseille <marseille.remi@gmail.com>
+ */
+class BackedEnumFilterTest extends DoctrineOrmFilterTestCase
```
Comment (soyuka) [inline src/Doctrine/Orm/Tests/Filter/BackedEnumFilterTest.php:23]: final

File: src/Doctrine/Orm/Filter/BackedEnumFilter.php (hunk 3 above)
Comment (mremi) [reply src/Doctrine/Orm/Filter/BackedEnumFilter.php:163]: looks like it is https://github.com/api-platform/core/blob/main/composer.json#L68

**Comment by soyuka** [comment]
thanks!

**Comment by mremi** [comment]
@soyuka when will be bumped the next release including it?

**Comment by dunglas** [comment]
@mremi during the API Platform Conference! It will be part of API Platform 3.4 and 4.0.

**Comment by mremi** [comment]
@dunglas nice! Thank you

**Comment by soyuka** [comment]
@mremi can you add a documentation for this?

**Comment by mremi** [comment]
> @mremi can you add a documentation for this?
You mean here https://github.com/api-platform/docs/blob/3.3/core/filters.md ?

**Pull request merged**